File: nebras-app/app/src/main/kotlin/com/nebras/mobile/feature/home/data/HomeRailBuilders.kt
package com.nebras.mobile.feature.home.data

import com.nebras.mobile.core.model.Content
import com.nebras.mobile.core.model.ContentType
import com.nebras.mobile.core.service.ContinueWatchingService
import com.nebras.mobile.core.service.InterestProfileService
import com.nebras.mobile.core.service.RecommendationService
import com.nebras.mobile.core.util.extractKeywords
import com.nebras.mobile.feature.home.model.HomeRail
import com.nebras.mobile.feature.home.model.HomeRailType
import com.nebras.mobile.feature.home.model.HomeSection

/**
 * يبني رفوف الصفحة الرئيسيّة من المحتوى والأقسام المحمّلة.
 *
 * **مهم:** ترتيب العناصر *داخل* أيّ قسم Firestore يبقى «الأقدم أولاً» عبر
 * `ContentOrdering` — الرفوف هنا تستخدم `createdAt` تنازلياً أو درجات
 * الشعبيّة فقط لعرض الاكتشاف في الصفحة الرئيسيّة.
 */
object HomeRailBuilders {

    private const val RAIL_LIMIT = 12

    fun allContentFlat(sections: Iterable<HomeSection>): List<Content> {
        val byId = LinkedHashMap<String, Content>()
        for (s in sections) {
            for (item in s.items) byId[item.id] = item
        }
        return byId.values.toList()
    }

    /**
     * تضفير حسب النوع (round-robin): يأخذ قائمة **مرتّبة مسبقاً** بأيّ معيار
     * (زمنيّ/شعبيّة/أفنيتي) ويُعيد توزيع المقاعد بين الأنواع بالتناوب مع
     * الحفاظ التامّ على الترتيب النسبيّ داخل كلّ نوع.
     *
     * لماذا: الرفوف كانت تستعمل معياراً أحاديّاً ثمّ قصّاً، فإن غلب نوعٌ على
     * رأس القائمة (مثلاً الكتب لأنّها الأكثر عدداً) اكتسح الرفّ كلّه.
     * التضفير يضمن ظهور بقيّة الأنواع دون كسر معيار الترتيب. إن لم يوجد
     * سوى نوع واحد يبقى الرفّ كما هو (تدهور سلس) — لا نُفرّغه أبداً.
     *
     * دالّة نقيّة حتميّة (بلا عشوائيّة/حالة) كي تبقى الاختبارات مستقرّة.
     */
    private fun interleaveByType(ranked: List<Content>, limit: Int): List<Content> {
        if (ranked.size <= 1) return ranked.take(limit)

        // طوابير لكلّ نوع تحافظ على ترتيب الإدخال (المعيار الأصليّ).
        val queues = LinkedHashMap<ContentType, MutableList<Content>>()
        val typeOrder = mutableListOf<ContentType>() // ترتيب أوّل ظهور (حتميّ).
        for (c in ranked) {
            queues.getOrPut(c.type) { mutableListOf() }.add(c)
            if (!typeOrder.contains(c.type)) typeOrder.add(c.type)
        }
        // نوع واحد فقط → لا فائدة من التضفير، أعِد الترتيب الأصليّ.
        if (typeOrder.size <= 1) return ranked.take(limit)

        val out = mutableListOf<Content>()
        val cursors = typeOrder.associateWith { 0 }.toMutableMap()
        var exhausted = false
        while (out.size < limit && !exhausted) {
            exhausted = true
            for (t in typeOrder) {
                val q = queues.getValue(t)
                val i = cursors.getValue(t)
                if (i < q.size) {
                    out.add(q[i])
                    cursors[t] = i + 1
                    exhausted = false
                    if (out.size >= limit) break
                }
            }
        }
        return out
    }

    fun buildNewestRail(pool: List<Content>, limit: Int = RAIL_LIMIT): List<Content> {
        // الأحدث يحكم الاختيار داخل كلّ نوع، والتضفير يمنع أن تكون الـ12 كتباً.
        val sorted = pool.sortedByDescending { it.createdAt }
        return interleaveByType(sorted, limit)
    }

    fun buildPopularRail(
        pool: List<Content>,
        popularIds: List<String> = emptyList(),
        limit: Int = RAIL_LIMIT,
    ): List<Content> {
        if (popularIds.isNotEmpty()) {
            val byId = pool.associateBy { it.id }
            val ordered = popularIds.mapNotNull { byId[it] }
            // نُضفّر مع الحفاظ على رتبة الشعبيّة داخل كلّ نوع.
            if (ordered.isNotEmpty()) return interleaveByType(ordered, limit)
        }
        val sorted = pool.sortedByDescending { it.viewCount }
        return interleaveByType(sorted, limit)
    }

    fun buildForYouRail(
        pool: List<Content>,
        excludeIds: Set<String> = emptySet(),
        limit: Int = RAIL_LIMIT,
    ): List<Content> {
        val filtered = pool.filterNot { excludeIds.contains(it.id) }
        val ranked = RecommendationService.rank(filtered, discoveryPercent = 15)
        // نُبقي ترتيب الأفنيتي لكن نكسر احتكار نوعٍ واحد للرفّ على مستوى
        // العرض — دون تعطيل التعلّم.
        return interleaveByType(ranked, limit)
    }

    data class SearchAffinityRail(val items: List<Content>, val subtitle: String?)

    fun buildSearchAffinityRail(
        pool: List<Content>,
        excludeIds: Set<String> = emptySet(),
        limit: Int = RAIL_LIMIT,
    ): SearchAffinityRail {
        val keywords = InterestProfileService.topKeywords(5)
        if (keywords.isEmpty()) return SearchAffinityRail(emptyList(), null)

        val scored = mutableListOf<Pair<Content, Double>>()
        for (item in pool) {
            if (excludeIds.contains(item.id)) continue
            val titleKeys = extractKeywords(item.title, 12)
            val descKeys = extractKeywords(item.description, 8)
            var score = 0.0
            for (k in keywords) {
                if (titleKeys.contains(k)) score += InterestProfileService.keywordAffinity(k) * 2
                if (descKeys.contains(k)) score += InterestProfileService.keywordAffinity(k)
            }
            if (score > 0) scored.add(item to score)
        }
        val items = scored.sortedByDescending { it.second }.take(limit).map { it.first }
        val subtitle = if (keywords.size >= 2) {
            "${keywords[0]}، ${keywords[1]}"
        } else {
            keywords.first()
        }
        return SearchAffinityRail(items, subtitle)
    }

    fun buildContinueRail(pool: List<Content>, limit: Int = RAIL_LIMIT): List<Content> {
        val entries = ContinueWatchingService.activeEntries()
        if (entries.isEmpty()) return emptyList()
        val byId = pool.associateBy { it.id }
        val out = mutableListOf<Content>()
        for (e in entries) {
            byId[e.contentId]?.let(out::add)
            if (out.size >= limit) break
        }
        return out
    }

    fun idsOf(items: Iterable<Content>): Set<String> = items.mapTo(mutableSetOf()) { it.id }

    /**
     * يبني الرفوف بالترتيب المحدَّد.
     *
     * [personalized]: عندما تكون `false` (ضيف بلا حساب حقيقيّ) لا نبني أيّ
     * رفّ يعتمد على سلوك المستخدم (لك / من اهتماماتك). الضيف يرى الأحدث
     * والأكثر مشاهدة فقط — التخصيص الكامل محجوز لأصحاب الحسابات.
     */
    fun buildAll(
        sections: List<HomeSection>,
        popularIds: List<String> = emptyList(),
        personalized: Boolean = false,
    ): List<HomeRail> {
        val pool = allContentFlat(sections)
        val rails = mutableListOf<HomeRail>()

        val newest = buildNewestRail(pool)
        rails.add(
            HomeRail(
                type = HomeRailType.NEWEST,
                title = "الجديد في نبراس",
                items = newest,
            ),
        )

        val popular = buildPopularRail(pool, popularIds)
        if (popular.isNotEmpty()) {
            rails.add(
                HomeRail(
                    type = HomeRailType.POPULAR,
                    title = "الأكثر مشاهدة",
                    items = popular,
                ),
            )
        }

        // رفّ «تابع المشاهدة» متاح للجميع (بمن فيهم الضيوف): التقدّم يُحفظ
        // محليّاً بصرف النظر عن وجود حساب، فمن غير المنطقيّ إخفاؤه عن الضيف
        // الذي بدأ مشاهدة فعليّة.
        val continueItems = buildContinueRail(pool)
        if (continueItems.isNotEmpty()) {
            rails.add(
                HomeRail(
                    type = HomeRailType.CONTINUE_WATCHING,
                    title = "تابع التصفّح",
                    items = continueItems,
                ),
            )
        }

        // باقي الرفوف الشخصيّة محجوزة لأصحاب الحسابات فقط.
        if (!personalized) return rails

        val excludeForLater = (idsOf(newest) + idsOf(popular)).toMutableSet()

        if (InterestProfileService.isLoaded && InterestProfileService.hasSignals) {
            val forYou = buildForYouRail(pool, excludeForLater)
            if (forYou.isNotEmpty()) {
                rails.add(
                    HomeRail(
                        type = HomeRailType.FOR_YOU,
                        title = "مقترَح لك",
                        items = forYou,
                    ),
                )
                excludeForLater.addAll(idsOf(forYou))
            }
        }

        val searchResult = buildSearchAffinityRail(pool, excludeForLater)
        if (searchResult.items.isNotEmpty()) {
            val title = searchResult.subtitle
                ?.let { "من اهتماماتك: $it" }
                ?: "من اهتماماتك في البحث"
            rails.add(
                HomeRail(
                    type = HomeRailType.SEARCH_AFFINITY,
                    title = title,
                    items = searchResult.items,
                ),
            )
            excludeForLater.addAll(idsOf(searchResult.items))
        }

        // رفّ «تصفّح الأقسام» السفليّ أُزيل عمداً — التنقّل بين الأقسام يتمّ
        // هرميّاً انطلاقاً من المحتوى (زرّ القسم) لا من شبكة سفليّة.

        return rails
    }
}

File: app-flutter/lib/features/home/services/home_rail_builders.dart
import 'package:easy_localization/easy_localization.dart';
import 'package:nebras_mobile_app/core/data/content_ordering.dart';
import 'package:nebras_mobile_app/core/services/continue_watching_service.dart';
import 'package:nebras_mobile_app/core/services/interest_profile_service.dart';
import 'package:nebras_mobile_app/core/services/recommendation_service.dart';
import 'package:nebras_mobile_app/core/utils/text_normalizer.dart';
import 'package:nebras_mobile_app/features/content/model/content_model.dart';
import 'package:nebras_mobile_app/features/home/data/home_section_model.dart';
import 'package:nebras_mobile_app/features/home/model/home_rail.dart';

/// يبني رفوف الصفحة الرئيسيّة من المحتوى والأقسام المحمّلة.
///
/// **مهم:** ترتيب العناصر *داخل* أيّ قسم Firestore يبقى عبر
/// [compareContentOldestFirst] — الرفوف هنا تستخدم `createdAt` تنازلياً
/// أو درجات الشعبيّة فقط لعرض الاكتشاف في الصفحة الرئيسيّة.
class HomeRailBuilders {
  HomeRailBuilders._();

  static const int _railLimit = 12;

  static List<Content> allContentFlat(Iterable<HomeSection> sections) {
    final byId = <String, Content>{};
    for (final s in sections) {
      for (final item in s.items) {
        byId[item.id] = item;
      }
    }
    return byId.values.toList();
  }

  static List<Content> buildNewestRail(List<Content> pool, {int limit = _railLimit}) {
    final copy = List<Content>.of(pool)
      ..sort((a, b) => b.createdAt.compareTo(a.createdAt));
    return copy.take(limit).toList(growable: false);
  }

  static List<Content> buildPopularRail(
    List<Content> pool, {
    List<String> popularIds = const [],
    int limit = _railLimit,
  }) {
    if (popularIds.isNotEmpty) {
      final byId = {for (final c in pool) c.id: c};
      final ordered = <Content>[];
      for (final id in popularIds) {
        final c = byId[id];
        if (c != null) ordered.add(c);
        if (ordered.length >= limit) break;
      }
      if (ordered.isNotEmpty) return ordered;
    }
    final copy = List<Content>.of(pool)
      ..sort((a, b) => b.viewCount.compareTo(a.viewCount));
    return copy.take(limit).toList(growable: false);
  }

  static List<Content> buildForYouRail(
    List<Content> pool, {
    Set<String> excludeIds = const {},
    int limit = _railLimit,
  }) {
    final filtered = pool.where((c) => !excludeIds.contains(c.id));
    final ranked = RecommendationService.instance.rank(
      filtered,
      discoveryPercent: 15,
    );
    return ranked.take(limit).toList(growable: false);
  }

  static ({List<Content> items, String? subtitle}) buildSearchAffinityRail(
    List<Content> pool, {
    Set<String> excludeIds = const {},
    int limit = _railLimit,
  }) {
    final profile = InterestProfileService.instance;
    final keywords = profile.topKeywords(limit: 5);
    if (keywords.isEmpty) {
      return (items: const <Content>[], subtitle: null);
    }

    final scored = <({Content c, double score})>[];
    for (final item in pool) {
      if (excludeIds.contains(item.id)) continue;
      final titleKeys = extractKeywords(item.title, maxTokens: 12);
      final descKeys = extractKeywords(item.description, maxTokens: 8);
      var score = 0.0;
      for (final k in keywords) {
        if (titleKeys.contains(k)) score += profile.keywordAffinity(k) * 2;
        if (descKeys.contains(k)) score += profile.keywordAffinity(k);
      }
      if (score > 0) scored.add((c: item, score: score));
    }
    scored.sort((a, b) => b.score.compareTo(a.score));
    final items = scored.map((e) => e.c).take(limit).toList(growable: false);
    final subtitle = keywords.length >= 2
        ? '${keywords[0]}، ${keywords[1]}'
        : keywords.first;
    return (items: items, subtitle: subtitle);
  }

  static Future<List<Content>> buildContinueRail(
    List<Content> pool, {
    int limit = _railLimit,
  }) async {
    final entries = await ContinueWatchingService.instance.activeEntries();
    if (entries.isEmpty) return const [];
    final byId = {for (final c in pool) c.id: c};
    final out = <Content>[];
    for (final e in entries) {
      final c = byId[e.contentId];
      if (c != null) out.add(c);
      if (out.length >= limit) break;
    }
    return out;
  }

  static Set<String> idsOf(Iterable<Content> items) =>
      items.map((c) => c.id).toSet();

  /// يبني الرفوف بالترتيب المحدّد في الوثائق.
  ///
  /// [personalized] : عندما تكون `false` (مستخدم ضيف بلا حساب حقيقيّ) لا
  /// نبني أيّ رفّ يعتمد على سلوك المستخدم (لك / من اهتماماتك / تابع
  /// التصفّح). الضيف يرى فقط الأحدث والأكثر مشاهدة — التطبيق لا يُكوّن له
  /// أيّ ملفّ اهتمامات. التخصيص الكامل محجوز لأصحاب الحسابات.
  static Future<List<HomeRail>> buildAll({
    required List<HomeSection> sections,
    required List<HomeSection> rankedSections,
    List<String> popularIds = const [],
    bool personalized = false,
  }) async {
    final pool = allContentFlat(sections);
    final rails = <HomeRail>[];

    final newest = buildNewestRail(pool);
    rails.add(
      HomeRail(
        type: HomeRailType.newest,
        title: 'home_rail_newest'.tr(),
        items: newest,
      ),
    );

    final popular = buildPopularRail(pool, popularIds: popularIds);
    if (popular.isNotEmpty) {
      rails.add(
        HomeRail(
          type: HomeRailType.popular,
          title: 'home_rail_popular'.tr(),
          items: popular,
        ),
      );
    }

    // الرفوف الشخصيّة محجوزة لأصحاب الحسابات فقط.
    if (!personalized) {
      return rails;
    }

    final excludeForLater = idsOf(newest)..addAll(idsOf(popular));

    final profile = InterestProfileService.instance;
    if (profile.isLoaded && profile.hasSignals) {
      final forYou = buildForYouRail(
        pool,
        excludeIds: excludeForLater,
      );
      if (forYou.isNotEmpty) {
        rails.add(
          HomeRail(
            type: HomeRailType.forYou,
            title: 'home_rail_for_you'.tr(),
            items: forYou,
          ),
        );
        excludeForLater.addAll(idsOf(forYou));
      }
    }

    final searchResult = buildSearchAffinityRail(
      pool,
      excludeIds: excludeForLater,
    );
    if (searchResult.items.isNotEmpty) {
      final title = searchResult.subtitle != null
          ? '${'home_rail_search_prefix'.tr()}: ${searchResult.subtitle}'
          : 'home_rail_search'.tr();
      rails.add(
        HomeRail(
          type: HomeRailType.searchAffinity,
          title: title,
          items: searchResult.items,
        ),
      );
      excludeForLater.addAll(idsOf(searchResult.items));
    }

    final continueItems = await buildContinueRail(pool);
    if (continueItems.isNotEmpty) {
      rails.add(
        HomeRail(
          type: HomeRailType.continueWatching,
          title: 'home_rail_continue'.tr(),
          items: continueItems,
        ),
      );
    }

    // رفّ «تصفّح الأقسام» السفليّ أُزيل عمداً — التنقّل بين الأقسام يتمّ
    // الآن هرميّاً انطلاقاً من المحتوى (زرّ القسم) لا من شبكة سفليّة.

    return rails;
  }
}
